### PR TITLE
Unit extraction pipeline to extract units from raw audio.

### PR DESCRIPTION
**Pull request by kauterry**
This PR does the following:
- Loads a `Wav2Vec2Model`, and implements a `Wav2Vec2LayerOutputModel` which uses its encoder_frontend & encoder to extraction features from a layer of the user's preference from xlsr2_v2.
- Runs K-Means clustering by loading a K-Means model (collection of centroids) to generate units from raw audio.
- Implements the `UnitExtractor` module which does the above.
- Introduce an `audio_to_units` directory to `scripts/m4t`.

TODO: Write a README for scripts/m4t/audio_to_units and add `m4t_audio_to_units` convenience command.

Test plan:

python scripts/m4t/audio_to_units/audio_to_units.py ~/eng_LJSpeech-1.1_0/LJ003-0001.wav
INFO:__main__:Running unit_extraction on the GPU.
Using the cached checkpoint of the model 'xlsr2_1b_v2'. Set `force=True` to download again.
Using the cached checkpoint of the model 'https://dl.fbaipublicfiles.com/seamlessM4T/models/unit_extraction/kmeans_10k.npy'. Set `force=True` to download again.
INFO:__main__:Converted to units: tensor([5884, 2411, 5280, 1803, 5932, 3421, 3992, 1867, 1867, 4783,  772, 2781,
        2103, 1839, 6327, 5579, 9121, 3375, 7563, 5932, 3421, 9014, 4783,  925,
        8528, 3254, 6466, 3254, 6394, 6394, 1838, 3380, 3407, 4085, 5682, 2386,
        5510, 4678, 7990, 5957, 5116,  111,  114, 7400, 4216, 9320, 4671, 2231,
        2231,  469,  469, 3796, 3382, 5687,  319, 7804, 1829, 3549, 1030,  337,
         337, 1600, 8975, 8834, 3214, 3397, 9403, 6297, 5956, 7350, 7350, 9917,
        6800, 8316,  296, 3661,  296, 7296, 7296, 3254, 2213,  856, 3294, 1677,
        6507, 2258, 1232, 2231, 9448, 9936, 6176, 3121, 5854, 7265, 6799,  179,
        6366, 9122, 5723, 5382, 5382, 8530, 1183, 4999, 6277, 2231,  469,  469,
         469,  469, 6551, 6336, 5424, 8379, 3214, 9173, 8073, 4082, 5152, 5152,
        8205, 3335, 8933, 1982, 4885, 8896, 8862, 5351, 4806, 1288, 6976, 4417,
        5415, 5415, 3602, 2412, 9351, 7754, 3540, 3454, 3454, 5127, 5127, 2229,
        5370, 7964,  386,  386, 5940, 5320, 5320, 5216, 2553, 3194,  984, 2399,
        3992, 8418, 8635, 8635, 2590, 1837, 8419, 8159, 5804,  730, 9488, 1364,
        1235, 4280,  492, 1000, 1000, 2631, 5484, 7183, 7183, 9226,  616,  352,
        8609, 7885, 7885, 8178, 1059,  493, 4009, 9583, 9583, 8881, 8881, 3163,
        9526, 5798,  830,  830,  360, 5934, 9884, 9884, 9980, 7943, 4940, 1330,
        6585, 6167, 2903, 6415, 5188, 6010, 5641, 6170, 6170, 7563, 6066, 2908,
        4133, 7991, 9939, 2487, 6007, 8689, 6363, 3570, 7606, 2289, 6105, 2121,
        3906, 3906, 3752, 3877, 9328, 7992, 4750, 4098, 5351, 5351, 6350, 1288,
        6350, 4148, 1154, 7087, 1103, 7787,  411, 5394, 2658, 5394,  366, 5771,
        8295, 4064, 7053, 5704, 6348, 9436, 9436, 4009, 3381, 5122, 8881, 2170,
        5829, 3163, 9324, 9324, 5829,  427, 9863, 8781, 8781, 6043, 5934, 7591,
        8487, 1450, 7140, 4452, 7450,  954, 4216,   66, 2258, 8151, 2231, 1326,
        8335, 5221, 5687, 2440,  319, 6818, 7731, 3549, 3549, 6545, 3549, 2284,
         337,  337, 9550, 2801, 5751, 1637, 8652, 1322, 3892, 4009, 7994, 8935,
        9699,  918, 5243, 4265, 1545, 5610, 4878, 5583, 1450, 8487, 1284, 4119,
        5030, 3430, 2079, 2079, 2254, 6515, 6740,  239, 6350, 4012, 3229, 4012,
        3229, 2927, 8890, 5905, 3570, 1299,  995, 6066, 1091, 5382, 5137, 5940,
        1757, 4859, 5804, 6243, 8113, 1859, 8018, 9055, 1544, 9055, 2777, 7857,
        9858, 4864, 1665, 3214, 8472, 4919, 9107, 1144, 9844, 4462, 7412, 4811,
        7631, 2939,  890, 9363,  890, 7556, 1697, 7601,  511, 1581, 7708, 6243,
        4807, 1829, 1030, 3549, 5705,  337, 7705, 9213, 7915, 6705, 4265, 9950,
        6366, 5589, 5382, 1144, 3095, 4961, 7412, 7103,  459, 7631,  744, 8034,
        9050, 3022, 9050, 3022, 3093, 8522, 4528, 5536, 9937, 2838, 7053, 5045,
        4069,  816, 3381, 3646, 3729, 5808, 6531, 9929, 1178,  376, 9033, 1669,
        1316, 2087, 6922, 7226, 7525, 4191, 3540, 3454, 7104, 7104, 4755, 8975,
        8497, 8975, 3591, 7972, 2495, 6366, 9122, 5382, 9247, 8845, 1144, 5106,
        4741, 8281, 9245, 2231,  469,  469,  469, 4584, 6336, 2122, 6336, 3214,
        6336, 4265, 6334], device='cuda:0')

### CODE REVIEW

File: scripts/m4t/audio_to_units/audio_to_units.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        default="xlsr2_1b_v2",
+    )
+    parser.add_argument(
+        "--vocoder_name", type=str, help="Vocoder name", default="vocoder_36langs"
+    )
+    parser.add_argument(
+        "--out_layer_number",
+        type=int,
+        help="Layer number of the feature extraction model to pull out features from.",
+        default=35,
```
Comment (mavlyutovr) [inline scripts/m4t/audio_to_units/audio_to_units.py:44]: nit: can we have this as a part of xlsr2_1b_v2.yaml (e.g as `default_out_layer`)? It will be great to not have to carry this constant across the repo

File: src/seamless_communication/models/unit_extraction/wav2vec2_layer_output.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            layer_padding_mask: Optional[Tensor],
+            num_layers: int,
+        ) -> None:
+            nonlocal w2v2_layer_output
+
+            if layer_idx == out_layer_idx:
+                w2v2_layer_output = layer_output
+
+        # TODO: Should pad for fp16?
+        _, _ = self.encoder(seqs, padding_mask, layer_output_hook)
```
Comment (mavlyutovr) [inline src/seamless_communication/models/unit_extraction/wav2vec2_layer_output.py:111]: I wonder if we can stop forward pass once the output layer is reached? (skip downstream layers)

File: src/seamless_communication/models/unit_extraction/wav2vec2_layer_output.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+            layer_output: Tensor,
+            layer_padding_mask: Optional[Tensor],
+            num_layers: int,
+        ) -> None:
+            nonlocal w2v2_layer_output
+
+            if layer_idx == out_layer_idx:
+                w2v2_layer_output = layer_output
+
+        # TODO: Should pad for fp16?
```
Comment (mavlyutovr) [inline src/seamless_communication/models/unit_extraction/wav2vec2_layer_output.py:110]: Can you clarify?

File: src/seamless_communication/models/unit_extraction/unit_extraction.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        out_layer_idx: int,
+    ) -> Tuple[List[Tensor], int]:
+        if isinstance(audio, str):
+            with Path(audio).open("rb") as fb:
+                block = MemoryBlock(fb.read())
+            decoded_audio = self.decode_audio(block)
+        else:
+            decoded_audio = {
+                "waveform": audio,
+                "sample_rate": 16000.0,
```
Comment (mavlyutovr) [inline src/seamless_communication/models/unit_extraction/unit_extraction.py:60]: nit: make 16000 and constant? Also, should it be an integer?

Comment (mavlyutovr) [reply src/seamless_communication/models/unit_extraction/unit_extraction.py:60]: Probably, even make sampling_rate as an optional argument to predict

File: src/seamless_communication/models/unit_extraction/unit_extraction.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        device: Device,
+        layer: int = 35,
+    ):
+        super().__init__()
+        self.model: Wav2Vec2LayerOutputModel = Translator.load_model_for_inference(
+            load_wav2vec2_layer_output_model, model_name_or_card, device, torch.float32
+        )
+        self.device = device
+        self.decode_audio = AudioDecoder(dtype=torch.float32, device=device)
+        self.collate = Collater(pad_idx=2, pad_to_multiple=2)
```
Comment (mavlyutovr) [inline src/seamless_communication/models/unit_extraction/unit_extraction.py:44]: Can we take pad_idx a constant? And ideally pull it from the model card?

File: scripts/m4t/audio_to_units/audio_to_units.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+from seamless_communication.models.vocoder import load_vocoder_model, Vocoder
+from itertools import groupby
+
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def main():
+    parser = argparse.ArgumentParser(
```
Comment (cndn) [inline scripts/m4t/audio_to_units/audio_to_units.py:21]: So the reason to convert back to audio is simply for verification? If we don't expect this as a common need maybe we could skip? Given we already abandoned earlier audio_synthesizer :)

File: src/seamless_communication/models/inference/translator.py [hunk 7]
```diff
@@ -74,10 +73,22 @@ def __init__(
             pad_idx=self.text_tokenizer.vocab_info.pad_idx, pad_to_multiple=2
         )
         # Load the vocoder.
-        self.vocoder: Vocoder = load_vocoder_model(vocoder_name_or_card, device=device)
-        self.vocoder.eval()
+        self.vocoder: Vocoder = self.load_model_for_inference(
+            load_vocoder_model, vocoder_name_or_card, device, torch.float32
```
Comment (cbalioglu) [inline src/seamless_communication/models/inference/translator.py:77]: Is there a specific reason why we use fp32 for Vocoder?

File: src/seamless_communication/models/inference/translator.py [hunk 8]
```diff
@@ -74,10 +73,22 @@ def __init__(
...
+        )
         self.sr = sample_rate
 
+    @staticmethod
+    def load_model_for_inference(
+        load_model_fn: Any,
```
Comment (cbalioglu) [inline src/seamless_communication/models/inference/translator.py:83]: nit: being pedantic here. Ideally this should be `Callable[[...], torch.nn.Module]`.

File: src/seamless_communication/models/inference/translator.py [hunk 9]
```diff
@@ -120,16 +131,6 @@ def get_modalities_from_task(self, task: Task) -> Tuple[Modality, Modality]:
...
-    def synthesize_speech(
-        self,
-        code: List[int],
-        lang: str,
-        speaker: Optional[int] = None,
-        dur_prediction: Optional[bool] = True,
-    ) -> Tuple[List[Tensor], int]:
-        return self.vocoder(code, lang, speaker, dur_prediction), self.sr
-
     @torch.no_grad()
```
Comment (cbalioglu) [inline src/seamless_communication/models/inference/translator.py:134]: Not part of this PR, but you should prefer `@torch.inference_mode()` here that completely disables autograd.

File: src/seamless_communication/models/inference/translator.py [hunk 10]
```diff
@@ -205,5 +206,5 @@ def predict(
             return text_out.sentences[0], None, None
         else:
             units = unit_out.units[:, 1:][0].cpu().numpy().tolist()
-            wav_out, sr_out = self.synthesize_speech(units, tgt_lang, spkr)
-            return text_out.sentences[0], wav_out, sr_out
+            wav_out = self.vocoder(units, tgt_lang, spkr, dur_prediction=True)
+            return text_out.sentences[0], wav_out, self.sr
```
Comment (cbalioglu) [inline src/seamless_communication/models/inference/translator.py:210]: I would prefer more descriptive names. What does `sr` mean?

File: src/seamless_communication/models/unit_extraction/kmeans.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+class KmeansModel(nn.Module):
+    def __init__(self, kmeans_uri: str, device: Device):
+        super().__init__()
+        km_path = download_manager.download_checkpoint(kmeans_uri, kmeans_uri)
+        km_model = np.load(km_path)
+        centroids_numpy = km_model.transpose()
+        centroids = torch.from_numpy(centroids_numpy)
+
+        self.centroids = nn.Parameter(centroids, requires_grad=False).to(device)
```
Comment (cbalioglu) [inline src/seamless_communication/models/unit_extraction/kmeans.py:23]: Why do we declare this as `Parameter` if we don't require grad? Same for centroid_norm.

File: src/seamless_communication/models/unit_extraction/kmeans.py [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+        self.centroid_norm = nn.Parameter(
+            (centroids**2).sum(0, keepdims=True), requires_grad=False
+        ).to(device)
```
Comment (cbalioglu) [inline src/seamless_communication/models/unit_extraction/kmeans.py:26]: Nit: call `to()` is redundant. `centroids` is already on `device`.

File: src/seamless_communication/models/unit_extraction/unit_extraction.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        self.kmeans_model = KmeansModel(kmeans_uri, device)
+
+    @torch.no_grad()
```
Comment (cbalioglu) [inline src/seamless_communication/models/unit_extraction/unit_extraction.py:47]: Same as above. Prefer `inference_mode()`.

File: src/seamless_communication/models/unit_extraction/unit_extraction.py [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+                block = MemoryBlock(fb.read())
+            decoded_audio = self.decode_audio(block)
+        else:
+            decoded_audio = {
+                "waveform": audio,
+                "sample_rate": 16000.0,
+                "format": -1,
+            }
+        src = self.collate(decoded_audio)["waveform"]
+        x = src["seqs"]
```
Comment (cbalioglu) [inline src/seamless_communication/models/unit_extraction/unit_extraction.py:64]: We still don't know whether this has parity with fairseq right?

File: src/seamless_communication/models/unit_extraction/unit_extraction.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+        x = F.layer_norm(x, x.shape)
```
Comment (cbalioglu) [inline src/seamless_communication/models/unit_extraction/unit_extraction.py:65]: Do we apply layer norm to the entire batch?? Typically it is applied only to the last dimension. Can you double check?

File: src/seamless_communication/models/unit_extraction/wav2vec2_layer_output.py (hunk 3 above)
Comment (cbalioglu) [reply src/seamless_communication/models/unit_extraction/wav2vec2_layer_output.py:110]: I think this is a remnant from fairseq2 code where we have a TODO item to evaluate the performance of fp16 with even vs. odd sequence lengths.

File: src/seamless_communication/models/unit_extraction/wav2vec2_layer_output.py [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+        num_codebooks=2,
+        num_codebook_entries=320,
+        codebook_sampling_temperature=(2, 0.1, 0.999995),
+        num_distractors=100,
+        logit_temp=0.1,
+        diversity_loss_weight=0.1,
+    )
+
+
+class Wav2Vec2LayerOutputModel(Wav2Vec2Model):
```
Comment (cbalioglu) [inline src/seamless_communication/models/unit_extraction/wav2vec2_layer_output.py:89]: I think this would be much cleaner if you just use the default `load_wav2vec2_model()` and instead of `Wav2Vec2LayerOutputModel`, define `TransformerEncoder` that takes a `TransformerFrontend` and `TransformerEncoder`. This would cut all the Builder stuff down below and don't "abuse" polymorphism here by subclassing Wav2Vec2Model.

File: src/seamless_communication/models/unit_extraction/unit_extraction.py (hunk 14 above)
Comment (kauterry) [reply src/seamless_communication/models/unit_extraction/unit_extraction.py:64]: yeah we don't, yet

File: src/seamless_communication/models/inference/translator.py (hunk 7 above)
Comment (cbalioglu) [reply src/seamless_communication/models/inference/translator.py:77]: Considering that we instantiate the model in fp16, I think it might make sense, but @cndn has much more experience with Vocoder, maybe there might be some culprit I am not aware of.

File: scripts/m4t/audio_to_units/audio_to_units.py (hunk 6 above)
Comment (kauterry) [reply scripts/m4t/audio_to_units/audio_to_units.py:21]: I'll just leave it as a convenience method in `UnitExtractor` then, which people can call and use. And remove it from here.

File: scripts/m4t/audio_to_units/audio_to_units.py (hunk 1 above)
Comment (kauterry) [reply scripts/m4t/audio_to_units/audio_to_units.py:44]: yeah sounds good

File: src/seamless_communication/models/inference/translator.py (hunk 10 above)
Comment (kauterry) [reply src/seamless_communication/models/inference/translator.py:210]: I'll change it to `sample_rate`

File: src/seamless_communication/models/inference/translator.py (hunk 8 above)
Comment (kauterry) [reply src/seamless_communication/models/inference/translator.py:83]: How about we go with `Callable[[Union[str, AssetCard], Device, DataType], nn.Module]`?

File: src/seamless_communication/models/inference/translator.py (hunk 7 above)
Comment (kauterry) [reply src/seamless_communication/models/inference/translator.py:77]: The vocoder checkpoint is in torch.float32, it's also evaluated in torch.float32, let me re-evaluate vocoders in torch.float16 and make sure they have parity before changing this.

File: src/seamless_communication/models/unit_extraction/wav2vec2_layer_output.py [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+)
+
+load_wav2vec2_layer_output_model = Wav2Vec2Loader(
+    asset_store,
+    download_manager,
+    create_wav2vec2_layer_output_model,
+    wav2vec2_archs,
+    # `weight_norm` used in `Wav2Vec2PositionEncoder` does not support meta
+    # initialization.
+    use_meta=False,
```
Comment (mavlyutovr) [inline src/seamless_communication/models/unit_extraction/wav2vec2_layer_output.py:175]: Need to remove that

File: src/seamless_communication/models/unit_extraction/kmeans.py (hunk 11 above)
Comment (kauterry) [reply src/seamless_communication/models/unit_extraction/kmeans.py:23]: will remove it

File: scripts/m4t/audio_to_units/audio_to_units.py (hunk 1 above)
Comment (kauterry) [reply scripts/m4t/audio_to_units/audio_to_units.py:44]: In fact, this `35` was determined empirically by @cndn, I think we should leave it to the user to use any layer number of their choice by suggesting a default.

File: src/seamless_communication/models/unit_extraction/wav2vec2_layer_output.py (hunk 2 above)
Comment (kauterry) [reply src/seamless_communication/models/unit_extraction/wav2vec2_layer_output.py:111]: To do that we have to add the logic in the `TransformerEncoder` in fairseq2, would @cbalioglu be open to that change?

class TransformerEncoder(nn.Module):

    def forward(self, seqs, padding_mask, layer_output_hook=None, stop_after_layer=None):

        for idx, layer in enumerate(self.layers): # assuming you have self.layers
            seqs, padding_mask = layer(seqs, padding_mask)
            if layer_output_hook:
                layer_output_hook(idx, seqs, padding_mask, len(self.layers))
            if stop_after_layer is not None and idx == stop_after_layer:
                break
                
        return seqs, padding_mask

File: src/seamless_communication/models/unit_extraction/wav2vec2_layer_output.py (hunk 16 above)
Comment (kauterry) [reply src/seamless_communication/models/unit_extraction/wav2vec2_layer_output.py:89]: Done!

File: src/seamless_communication/models/unit_extraction/unit_extraction.py (hunk 15 above)
Comment (kauterry) [reply src/seamless_communication/models/unit_extraction/unit_extraction.py]: Checked, no disparity with this operation wrt fairseq.

**Review by mavlyutovr — APPROVED** [review]

File: src/seamless_communication/models/unit_extraction/unit_extraction.py (hunk 14 above)
Comment (kauterry) [reply src/seamless_communication/models/unit_extraction/unit_extraction.py:68]: It doesn't have parity, but the output features are almost the same, and units are the same.
